zingolib/src/wallet/transaction_record.rs

```diff
@@ -0,0 +1,343 @@
+            }
+        }
+    }
+}
+//get
+impl TransactionRecord {
+    pub fn get_transparent_value_spent(&self) -> u64 {
+        self.total_transparent_value_spent
+    }
+    pub fn get_transaction_fee(&self) -> Result<u64, ZingoLibError> {
+        let outputted = self.value_outgoing() + self.total_change_returned();
+        if self.total_value_spent() >= outputted {
+            Ok(self.total_value_spent() - outputted)
+        } else {
+            ZingoLibError::MetadataUnderflow(format!(
+                "for txid {} with status {}: spent {}, outgoing {}, returned change {} \n {:?}",
+                self.txid,
+                self.status,
+                self.total_value_spent(),
+                self.value_outgoing(),
+                self.total_change_returned(),
+                self,
+            ))
+            .handle()
+        }
+    }
+
+    // TODO: This is incorrect in the edge case where where we have a send-to-self with
+    // no text memo and 0-value fee
+    pub fn is_outgoing_transaction(&self) -> bool {
+        (!self.outgoing_tx_data.is_empty()) || self.total_value_spent() != 0
+    }
+    pub fn is_incoming_transaction(&self) -> bool {
+        self.sapling_notes.iter().any(|note| !note.is_change())
+            || self.orchard_notes.iter().any(|note| !note.is_change())
+            || !self.transparent_notes.is_empty()
+    }
+    pub fn net_spent(&self) -> u64 {
+        assert!(self.is_outgoing_transaction());
+        self.total_value_spent() - self.total_change_returned()
+    }
+    fn pool_change_returned<D: DomainWalletExt>(&self) -> u64
+    where
+        <D as Domain>::Note: PartialEq + Clone,
+        <D as Domain>::Recipient: traits::Recipient,
+    {
+        D::sum_pool_change(self)
+    }
+
+    pub fn pool_value_received<D: DomainWalletExt>(&self) -> u64
+    where
+        <D as Domain>::Note: PartialEq + Clone,
+        <D as Domain>::Recipient: traits::Recipient,
+    {
+        D::to_notes_vec(self)
+            .iter()
+            .map(|note_and_metadata| note_and_metadata.value())
+            .sum()
+    }
+    pub fn total_change_returned(&self) -> u64 {
+        self.pool_change_returned::<SaplingDomain<ChainType>>()
+            + self.pool_change_returned::<OrchardDomain>()
+    }
+    pub fn total_value_received(&self) -> u64 {
+        self.pool_value_received::<OrchardDomain>()
+            + self.pool_value_received::<SaplingDomain<ChainType>>()
+            + self
+                .transparent_notes
+                .iter()
+                .map(|utxo| utxo.value)
+                .sum::<u64>()
+    }
+    pub fn total_value_spent(&self) -> u64 {
+        self.value_spent_by_pool().iter().sum()
+    }
+
+    pub fn value_outgoing(&self) -> u64 {
+        self.outgoing_tx_data
+            .iter()
+            .fold(0, |running_total, tx_data| tx_data.value + running_total)
+    }
+
+    pub fn value_spent_by_pool(&self) -> [u64; 3] {
```

Comment:
This contract may be hard to maintain going forward. I note that 'sprout' isn't represented, so over time as pools may be deprecated in the future, how can a user of this API confidently know which pool is at which index?
IMO it would be better to return a struct where each pool is named.

Reply:
agree, it would be.